operators/advanced_network/python/tune_system.py

```diff
@@ -0,0 +1,629 @@
+
+def check_max_payload_size():
+    """
+    Checks the Maximum Payload Size (MPS) of NVIDIA Ethernet controllers
+    from the DevCtl section and ensures it is set to 256 bytes.
+    """
+    try:
+        names, addrs = get_nic_info()
+        for idx in range(len(addrs)):
+            pci_address = addrs[idx]
+            name = names[idx]
+
+            # Query detailed device information using lspci -vv
+            mps_result = subprocess.run(
+                ["lspci", "-vv", "-s", pci_address], capture_output=True, text=True, check=True
+            )
+
+            # Parse MaxPayload information from the DevCtl section
+            lines = mps_result.stdout.splitlines()
+            devctl_found = False
+            for i, line in enumerate(lines):
+                if "DevCtl:" in line:
+                    devctl_found = True
+                    base_indent = len(line) - len(line.lstrip())  # Indentation level of DevCtl
+                    # Look for MaxPayload in subsequent indented lines
+                    for j in range(i + 1, len(lines)):
+                        current_indent = len(lines[j]) - len(lines[j].lstrip())
+                        if current_indent <= base_indent:  # Stop if indentation decreases or ends
+                            break
+                        if "MaxPayload" in lines[j]:
+                            # Extract the actual MaxPayload value from the line
+                            payload_info = lines[j].strip()
+                            max_payload_value = int(
+                                payload_info.split("MaxPayload")[1].split("bytes")[0].strip()
+                            )
+                            if max_payload_value == 256:
+                                logging.info(
+                                    f"{name}/{pci_address}: PCIe Max Payload Size is correctly set to 256 bytes."
+                                )
+                            else:
+                                logging.warning(
+                                    f"{name}/{pci_address}: PCIe Max Payload Size is not set to 256 bytes. Found: {max_payload_value} bytes."
+                                )
+                            break
+                    else:
+                        logging.error(
+                            f"{name}/{pci_address}: Unable to find MaxPayload information under DevCtl."
+                        )
+                    break
+
+            if not devctl_found:
+                logging.error(f"{name}/{pci_address}: DevCtl section not found.")
+
+    except FileNotFoundError:
+        logging.error("The required tools (lspci) are not available on this system.")
+        sys.exit(1)
+    except subprocess.CalledProcessError as e:
+        logging.error(f"Error while checking Max Payload Size: {e}")
+        sys.exit(1)
+
+
+def check_hugepages():
+    """
+    Checks if hugepages are allocated and ensures that the total allocated hugepage memory
+    is at least 500 MB.
```

Comment:
would we be happy if it's only 64kb size? Thought the size mattered too not just total

Reply:
That's true. Like the other things, the answer is that it depends. Specifically some systems may only want to use certain sizes (maybe they already use it for another application), and they don't want the bigger ones. I'm ok with just saying "hugepages are enabled/used" as the metric rather than trying to validate a certain size.

Reply:
Here is my thought:
1. when running the script by itself outside the context of applications, we should either
  a. have arbitrary rules, or
  b. don't do it at all (here just checking hugepages are created is fine, for other stuff like max payload size, mrrs, mtu etc just skip)
2. when running at startup in the app, we should check what specific config is done, and that might help augment these checks with more specific heuristics. And if users still don't like them, it's just warnings not errors, and they can ignore them, or we can have a flag to skip them.

I just would like to be consistent. IMO I'd rather ask "too much" in the vanilla run of the script to recommend optimal settings, since you could argue that every single one of these checks also dependson the need (you might be fine even without a lot of these settings), but maybe that's not reasonable?

Reply:
>for other stuff like max payload size, mrrs, mtu etc just skip

I didn't follow this part. Why would we not want to warn if their system is not set up optimally? Hugepages can be considered some level of optimality if any size is enabled and not just the largest.

> I just would like to be consistent. IMO I'd rather ask "too much" in the vanilla run of the script to recommend optimal settings, since you could argue that every single one of these checks also dependson the need (you might be fine even without a lot of these settings), but maybe that's not reasonable?

This is what Rivermax does. When you start it you typically get a whole bunch of warnings about things you should update. I don't see any issues warning them even if they never plan to do it since it seems like being more aggressive is better.

Reply:
no that sounds great to me, we're on the same page there, I was just thinking that we could do the same with huge pages, and say the optimal one is probably large huge pages than 64k (especially since you have larger sizes available like 1gb here) and if they don't agree they can ignore it.

Reply:
We could do that. Detecting hugepage sizes when they're not enabled is a real pain from what I recall. You have to check CPU flags and it's not the same from ARM to x86. When they're enabled you can just check the sys fs and walk through all the sizes of the directories

Reply:
Can we add that as a later step to get this merged?

Reply:
sure. and I meant when enabled already, so just making a check on the page size enabled (and not just the sum of them)